### PR TITLE
Use new Poetry installer and default to 1.2.0

### PR DESCRIPTION
**Pull request by davegaeddert**
The current (old) installer is totally broken, as far as I can tell. I don't think passing `GET_POETRY_IGNORE_DEPRECATION` works because it starts with `GET_`?

```
-----> Install Poetry
       Retrieving Poetry metadata
       
       This installer is deprecated, and cannot install Poetry 1.2.0a1 or newer.
       Additionally, Poetry installations created by this installer cannot `self update` to 1.2.0a1 or later.
       You should migrate to https://install.python-poetry.org instead. Instructions are available at https://python-poetry.org/docs/#installation.
       
       This installer will now exit. If you understand the above warning and wish to proceed anyway, set GET_POETRY_IGNORE_DEPRECATION=1 in the environment and run this installer again.
```

I see #39 which mentions some errors, but I didn't get any here. Maybe those have been fixed since then?

### CODE REVIEW

**Comment by jmduke** [comment]
(I can confirm that the current installer is broken on my end as well, so I appreciate the speed with which you got this PR up.)

**Comment by buckmaxwell** [comment]
Thank you! Yes, we need this ASAP

UPDATE: can confirm this works. we just switched our staging buildpack to use.

**Comment by zyv** [comment]
Hi folks, thank you very much for the PR! Could you please confirm which Heroku stack are you on? The problem in #38 / #39 was that we were still using the old Ubuntu 20.04 stack, but we didn't test again with Ubuntu 22.04.

**Comment by zyv** [comment]
@marns93 I wonder why merge checks don't seem to be working here? I'm also confused at to whether `/rebase` worked correctly in #39...

**Comment by 8bitcurser** [comment]
Thanks @davegaeddert ! I forked with your changes and it worked! Im using the most recent stack 22.04 @zyv

**Comment by davegaeddert** [comment]
> Could you please confirm which Heroku stack are you on?

Ah, ok that makes sense. I'm on heroku-22. What do you want to do about older stacks? Right now I'm guessing *all* stacks are broken, so if something non-trivial has to happen with the older ones, maybe it's ok to merge the 22 support first?

> @marns93 I wonder why merge checks don't seem to be working here? I'm also confused at to whether `/rebase` worked correctly in #39...

I pushed another commit — shellcheck didn't like my single quotes.

**Comment by ulgens** [comment]
@zyv 

>  I wonder why merge checks don't seem to be working here? 

CleanShot 2022-08-31 at 21 29 25@2x

I'm not sure there is bot in the repo executing `/rebase` command.

**Comment by 8bitcurser** [comment]
If anyone runs into this set poetry version to `1.1.15` if not the build is going to fail

**Comment by camerondavison** [comment]
The fork worked for us. We are on heroku-20

**Comment by jonahkagan** [comment]
The fork worked for us as well but only when setting `POETRY_VERSION=1.1.15` in the app config vars, which is maybe what @8bit-curser was recommending. We're on heroku-18.

**Comment by cfal** [comment]
when specifying `POETRY_VERSION=1.2.0` even with the forked version, it fails when exporting requirements.txt:

```
The currently activated Python version 3.8.10 is not supported by the project (3.9.11).
Trying to find and use a compatible version. 
Poetry was unable to find a compatible version. If you have one, you can explicitly use it via the "env use" command.
```

i guess that means that the default python version on our heroku stack (before the heroku python buildpack is executed) defaults to 3.8? would `poetry env use python<version>` (where version would be dependent on the stack?) as suggested at https://github.com/python-poetry/poetry/issues/4478 before exporting requirements fix the problem?

**Comment by marns93** [comment]
> @marns93 I wonder why merge checks don't seem to be working here? I'm also confused at to whether `/rebase` worked correctly in #39...

Our `merge-checks ` are running if the branch is prefixed with `feature/`. We need to think about, but perhaps we can just exclude `master` and trigger on all other branches.

Let's discuss the rebase stuff later on.

**Comment by kentuck13** [comment]
> when specifying `POETRY_VERSION=1.2.0` even with the forked version, it fails when exporting requirements.txt:
> 
> ```
> The currently activated Python version 3.8.10 is not supported by the project (3.9.11).
> Trying to find and use a compatible version. 
> Poetry was unable to find a compatible version. If you have one, you can explicitly use it via the "env use" command.
> ```
> 
> i guess that means that the default python version on our heroku stack (before the heroku python buildpack is executed) defaults to 3.8? would `poetry env use python<version>` (where version would be dependent on the stack?) as suggested at [python-poetry/poetry#4478](https://github.com/python-poetry/poetry/issues/4478) before exporting requirements fix the problem?

I have the same problem

**Comment by zyv** [comment]
> > @marns93 I wonder why merge checks don't seem to be working here? I'm also confused at to whether `/rebase` worked correctly in #39...
> 
> Our `merge-checks ` are running if the branch is prefixed with `feature/`. We need to think about, but perhaps we can just exclude `master` and trigger on all other branches.

Ah, I see :) thanks - maybe we should leave it as is, makes it nice to enforce the `feature/` rule, otherwise one'd be able to merge everything and anything.

**Comment by dustinchilson** [comment]
> > when specifying `POETRY_VERSION=1.2.0` even with the forked version, it fails when exporting requirements.txt:
> > ```
> > The currently activated Python version 3.8.10 is not supported by the project (3.9.11).
> > Trying to find and use a compatible version. 
> > Poetry was unable to find a compatible version. If you have one, you can explicitly use it via the "env use" command.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > i guess that means that the default python version on our heroku stack (before the heroku python buildpack is executed) defaults to 3.8? would `poetry env use python<version>` (where version would be dependent on the stack?) as suggested at [python-poetry/poetry#4478](https://github.com/python-poetry/poetry/issues/4478) before exporting requirements fix the problem?
> 
> I have the same problem

@cfal @kentuck13 I was able to solve the compatible version issue on the heroku-22 stack and python 3.10.6

https://github.com/CavalloSolutions/python-poetry-buildpack which is a fork of this PR

**Comment by ekersten** [comment]
@dustinchilson 's solution worked fine on our company's dokku server

**Comment by davegaeddert** [comment]
I merged @dustinchilson's addition into this PR (thanks @dustinchilson!)

**Comment by cfal** [comment]
@davegaeddert @dustinchilson the PR breaks support for 1.1.14 and lower:

```
-----> Export requirements.txt from Poetry
  ValueError
  Setting virtualenvs.prefer-active-python does not exist
  at ~/.local/share/pypoetry/venv/lib/python3.8/site-packages/poetry/console/commands/config.py:248 in handle
      244│                 raise ValueError("You must pass exactly 1 value")
      245│ 
      246│             return 0
      247│ 
    → 248│         raise ValueError("Setting {} does not exist".format(self.argument("key")))
      249│ 
      250│     def _handle_single_value(self, source, key, callbacks, values):
      251│         validator, normalizer, _ = callbacks
      252│ 
 !     Push rejected, failed to compile Python Poetry app.
 !     Push failed
```

**Comment by dustinchilson** [comment]
I was about to note I think my fix only works for poetry >=1.2.0

**Comment by davegaeddert** [comment]
Might have to do something better but I pushed a quick fix for now (should only run if Poetry 1.2.x). @cfal does 1.1.14 work now?

**Comment by cfal** [comment]
@davegaeddert yep, thanks for the quick fix

**Comment by davegaeddert** [comment]
@zyv I think the [lint-and-test stuff should pass](https://github.com/davegaeddert/python-poetry-buildpack/runs/8138273620?check_suite_focus=true) now if it runs. Anything else that needs to happen here? Do I need to make a new branch/PR under `feature/` so the merge checks run, or are we close to merging this PR? Thanks!

File: bin/compile [hunk 1]
```diff
@@ -63,7 +66,7 @@ fi
 
 # pip can't handle vcs & editable dependencies when requiring hashes (https://github.com/pypa/pip/issues/4995)
 # poetry exports local dependencies as editable by default (https://github.com/python-poetry/poetry/issues/897)
-poetry export -f "$REQUIREMENTS_FILE" "${EXPORT_PARAMETERS[@]}" > "$REQUIREMENTS_FILE.orig"
+poetry export -f requirements.txt "${EXPORT_PARAMETERS[@]}" -o "$REQUIREMENTS_FILE.orig" | indent
```
Comment (kavdev) [inline bin/compile:70]: Why not keep `"$REQUIREMENTS_FILE"` here?

Comment (zyv) [reply bin/compile:70]: Right, why this change? Then you don't need to extend poetry mock. This is the reason why I didn't use `-o` even though I knew about it.

File: bin/compile [hunk 2]
```diff
@@ -36,16 +36,19 @@ log "Generate requirements.txt with Poetry"
...
 
-curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py \
+curl -sSL https://install.python-poetry.org \
     | sed -e 's/allowed_executables = \["python", "python3"\]/allowed_executables = ["python3"]/' \
     | python3 \
     | indent
 
-PATH_TO_POETRY_ENV="$HOME/.poetry/env"
+log "Add poetry to the PATH"
+export PATH="/app/.local/bin:$PATH"
```
Comment (zyv) [inline bin/compile:45]: Why this change? Is it no longer working in new Poetry?

File: bin/compile [hunk 3]
```diff
@@ -36,16 +36,19 @@ log "Generate requirements.txt with Poetry"
...
 
-if [ -f "$PATH_TO_POETRY_ENV" ] ; then
-    # shellcheck source=/dev/null
-    . "$PATH_TO_POETRY_ENV"
+if [ "${POETRY_VERSION:0:3}" = "1.2" ] ; then
+    # Allow `poetry export` to work even if we aren't using the
+    # Python version that the project will ultimately run in
+    log "Set virtualenvs.prefer-active-python to true"
+    poetry config virtualenvs.prefer-active-python true | indent
```
Comment (zyv) [inline bin/compile:51]: Doesn't feel too great, but probably this is indeed the way to go. We are testing this on our infra right now.

File: bin/compile (hunk 1 above)
Comment (davegaeddert) [reply bin/compile:70]: It made it look like you could put any path there, when in reality the only option is the exact string `requirements.txt` as an output format. https://python-poetry.org/docs/cli/#options-12

Comment (zyv) [reply bin/compile:70]: Oh, that's actually a good point. Thank you! And how about the `-o` part?

Comment (davegaeddert) [reply bin/compile:70]: The issue I had with not using `-o` was that it put a “Creating virtualenv…” string at the top of every the output file which came through stdout: https://github.com/python-poetry/poetry/issues/6275

So if you just pipe the output, pip fails later when it tries to install the first line of requirements.txt.

Comment (kavdev) [reply bin/compile:69]: @davegaeddert it looks like it's just a convenience variable in case the file name changes from `requirements.txt` in the future. If it changes, it only changes in one line of the code.

It's set [on line 54](https://github.com/moneymeets/python-poetry-buildpack/pull/44/files/34a423d6ae9074e42f817829ba87853040970d46#diff-a64a7d9b82578ca3053416d56631a61e0ceabed87619415663030e3842429e05R54)

File: bin/compile (hunk 2 above)
Comment (davegaeddert) [reply bin/compile:45]: It was not working with 1.2.0. When I removed it though I wasn’t thinking about past versions… my guess is that the new installer puts them all in the same place now? Sounds Iike people have been testing older versions and not had an issue with it being gone.

File: bin/compile (hunk 1 above)
Comment (kavdev) [reply bin/compile:69]: I'm not sure about the merits of `-o` vs `>`, so I'll default to @zyv on that. I'm curious though.

Comment (zyv) [reply bin/compile:70]: I see, so they have added output that isn't going to stderr like before, but mixes with stdout. Extending the mock is annoying, but it seems that's indeed the best way. Thanks!

File: bin/compile (hunk 3 above)
Comment (davegaeddert) [reply bin/compile:51]: Agreed! Could probably implement it in different ways, but I’m not sure what else to do besides “use the Python currently available” and I don’t *think* the Python version you’re running affects the `poetry export` output? Didn’t totally confirm that myself though.

**Comment by felix11h** [comment]
Thank you for the contributions, @davegaeddert @kavdev @dustinchilson and others!

I have spent some time yesterday trying to understand what's going on, specifically with the mismatching of Python versions and the fix to add `poetry config virtualenvs.prefer-active-python true`.

In general, I'm ok with adding this line if this gets everyone moving forward. I could test and confirm that this works for us as well. It would be important to me, however, to generate some understanding of what is happening here, as my personal conclusion is that setting `prefer-active-python true` doesn't actually have much to do with the problem and only works because this command is _likely broken_.
<br/> 

### Main issue

Before, in Poetry 1.1, the command `poetry export` does not check or require that a Python version that matches the Python version required in the `pyproject.toml` is available. A new virtualenv is not created.

```console
# with python = "3.10.5" in pyproject.toml 
$ poetry --version
Poetry version 1.1.15
$ python -version
Python 3.10.4
$ poetry export -f requirements.txt -o requirements.txt
```

Now, in Poetry 1.2, this is required. The command `poetry export` fails, if it can't find a Python version that matches the required version. In our case, for example, 3.10.5. This is the main issue here, as on Heroku, the Python version initially available will depend on the Heroku version and will very likely not match the exact project version.

```console
# with python = "3.10.5" in pyproject.toml 
$ poetry --version
Poetry (version 1.2.0)
$ python -version
Python 3.10.4
$ poetry export -f requirements.txt -o requirements.txt
The currently activated Python version 3.10.4 is not supported by the project (3.10.5).
Trying to find and use a compatible version. 

Poetry was unable to find a compatible version. If you have one, you can explicitly use it via the "env use" command.
```
<br/>

### Prefer Virtualenv Fix

So how does `virtualenvs.prefer-active-python true` solve this? It really doesn't, as even with this setting, the correctly matching Python version will simply not be available. Yet, it somehow makes things work for us. In the Poetry documentation, this feature is marked as [experimental](https://python-poetry.org/docs/master/configuration/#virtualenvsprefer-active-python-experimental). I don't believe it's working correctly.


```console
# with python = "3.10.5" in pyproject.toml 
$ poetry --version
Poetry (version 1.2.0)
$ poetry config virtualenvs.prefer-active-python true --local
$ poetry export -f requirements.txt -o requirements.txt
The currently activated Python version 3.10.4 is not supported by the project (3.10.5).
Trying to find and use a compatible version. 

Creating virtualenv test-Slf4yrPO-py3.10 in /Users/felixhoffmann/Library/Caches/pypoetry/virtualenvs
```

The export has indeed worked correctly. We also see however that, unlike the behaviour in Poetry 1.1, a new virtualenv was created. If we look at that new virtualenv, we see

```console
$ poetry env info
Virtualenv
Python:         3.10.4
Implementation: CPython
Path:           /Users/felixhoffmann/Library/Caches/pypoetry/virtualenvs/test-Slf4yrPO-py3.10
Executable:     /Users/felixhoffmann/Library/Caches/pypoetry/virtualenvs/test-Slf4yrPO-py3.10/bin/python
Valid:          True
```

So in our 3.10.5 project, a virtualenv with Python version 3.10.4 was created. That doesn't seem quite right! To top things off, running again

```console
$ poetry export -f requirements.txt -o requirements.txt
Current Python version (3.10.4) is not allowed by the project (3.10.5).
Please change python executable via the "env use" command.
```

now yields an error. 

In summary, setting `virtualenvs.prefer-active-python true` allows to run exactly one Poetry command if a virtualenv isn't already present, as it will _incorrectly_ build this virtualenv, even when the Python versions aren't matching.

Do we want to make use of this? Do we want to create a broken virtualenv in our Dynos? I don't have a great feeling about this. However, I also don't have a great solution otherwise. What we would need, it seems, is to be able to run `poetry export` without checks and creation of a new virtualenv as in Poetry 1.1.

Any more thoughts on this? If there's consensus that we want to use this hack(?) to make things work for now, I'm happy to take over merging this PR, but it would be good to get feedback if this is really acceptable, @zyv @davegaeddert @kavdev.

**Comment by zyv** [comment]
@felix11h how about we just set https://python-poetry.org/docs/configuration/#virtualenvscreate to `false` ?

All we need is to be able to run Poetry in the context of the current project with whatever system Python version is available, and we only actually need commands that don't have anything to do with virtualenv management / installing dependencies / ... etc. This way we will avoid creating broken virtualenv with new Poetry versions and still be able to run the `export` command with system Python, whichever it is.

I find it very questionable that `export` command (not being a venv management command like `add`, `install` and so on) actually implicitly creates a virtualenv in the first place, but that's Poetry's call.

File: bin/compile (hunk 2 above)
Comment (zyv) [reply bin/compile:45]: @felix11h did you have a look how this works in 1.1 / 1.2?

File: bin/compile (hunk 3 above)
Comment (zyv) [reply bin/compile:51]: No, it shouldn't. This is why I'm irritated that since Poetry 1.2 they decided to enforce the selection of the virtualenv to be able to run `export` in the first place. But maybe this has something to do with `export` now being offloaded into a plugin, and they didn't really think about the consequences.

File: test/utils [hunk 4]
```diff
@@ -58,6 +58,10 @@ function curl() {
 export -f curl
 
 function poetry() {
+    if [ "$1" = "export" ] ; then
+        # Fake creating the export file
+        touch requirements.txt.orig
```
Comment (zyv) [inline test/utils]:
```suggestion
        # mock creating the export file
        last="$#"
        touch "${!last}"
```

Would this work @felix11h ?

**Comment by felix11h** [comment]
> @felix11h how about we just set https://python-poetry.org/docs/configuration/#virtualenvscreate to false ?

We can do this, in combination with `poetry config virtualenvs.prefer-active-python true` this works (but not without prefer active). 

It does yield some rather paradoxical results, however:

```console
$ poetry --version
Poetry (version 1.2.0)
$ poetry config virtualenvs.prefer-active-python true --local
$ poetry config virtualenvs.create false --local
$ poetry run python --version
The currently activated Python version 3.10.4 is not supported by the project (3.10.5).
Trying to find and use a compatible version. 
Skipping virtualenv creation, as specified in config file.

Python 3.10.4
```

**Comment by zyv** [comment]
> > @felix11h how about we just set https://python-poetry.org/docs/configuration/#virtualenvscreate to false ?
> 
> We can do this, in combination with `poetry config virtualenvs.prefer-active-python true` this works (but not without prefer active).

Yes, sorry for being unclear. I meant to set it **in addition to** `virtualenvs.prefer-active-python` to prevent broken venv creation, not just use it alone.

> It does yield some rather paradoxical results, however:

In which way are they paradoxical to you?

I thought that this is exactly what we want to achieve: force Poetry to use the Python version that is available for export, whatever it is and ignore whatever is required by the project. We don't use Poetry for anything other that requirements export, and it should work with whatever Python version that is still supported by Poetry itself.

**Comment by felix11h** [comment]
> I thought that this is exactly what we want to achieve: force Poetry to use the Python version that is available for export, whatever it is and ignore whatever is required by the project. We don't use Poetry for anything other that requirements export, and it should work with whatever Python version that is still supported by Poetry itself.

Yes, I agree, and it is definitely an improvement over the creation of the virtualenv. It is paradoxical to me in that 3.10.4 is reported as an incompatible version, a matching version is looked for, not found, and – if and only if prefer active is enabled – the command still works ... using Python Version 3.10.4, which was reported as incompatible. 

I'm really not sure that this is the intended use of prefer active, at least judging from the documentation: 

> Use currently activated Python version to create a new virtual environment. If set to false, Python version used during Poetry installation is used.

I agree that it should be ok for now, but I'm guessing that prefer active will eventually be fixed, in turn breaking our buildpack. Would be happy to be wrong about this!

---

EDIT: Some more context from [release notes](https://python-poetry.org/blog/announcing-poetry-1.2.0/) of Poetry 1.2

> Due to refactoring required to enable plugin support and alternative install methods, Poetry lost the ability to detect the currently activated Python (aka the current python3 command in the $PATH), as selected by tools like pyenv or update-alternatives.

> To ease the workflow of those using such tools, a new experimental [virtualenvs.prefer-active-python](https://python-poetry.org/docs/configuration/#virtualenvsprefer-active-python-experimental) setting has been introduced. If this is set to true, Poetry will attempt to detect the currently active Python interpreter when creating a new environment. This new method should function regardless of the install method used.

File: test/utils (hunk 4 above)
Comment (felix11h) [reply test/utils]: I think so! See CI status in https://github.com/moneymeets/python-poetry-buildpack/commit/a7fae83d25b042007638a3e4bce79deabfa30c10.

File: bin/compile (hunk 2 above)
Comment (felix11h) [reply bin/compile:45]: I can confirm that this works for us in 1.1 and 1.2. From a quick glance comparing the old and new installer script (note that also Poetry 1.1 needs to use the new script), this seems to be one of the changes when `$POETRY_HOME` isn't set.

File: test/utils (hunk 4 above)
Comment (felix11h) [reply test/utils]: Could you add this, @davegaeddert? Thank you!

**Comment by felix11h** [comment]
Assuming we want to move forward with 

```
poetry config virtualenvs.prefer-active-python true
poetry config virtualenvs.create false
```

@davegaeddert, could you add the second line? I will disable our merge-checks for this PR, but it would be great if you could still squash your commits into a single, or a couple of logically separated, commit(s). Thank you!

**Comment by zyv** [comment]
@davegaeddert @felix11h just to make it easier for everyone, I suggest we open a new PR - I can do all the changes discussed up to now.

**Comment by davegaeddert** [comment]
@zyv whatever you want to do sounds good to me! I’m on an iPad for the next few hours, so fixing the tests and especially squashing commits are a little problematic…

File: bin/compile [hunk 5]
```diff
@@ -36,16 +36,20 @@ log "Generate requirements.txt with Poetry"
 
 log "Install Poetry"
 
-curl -sSL https://raw.githubusercontent.com/python-poetry/poetry/master/get-poetry.py \
+curl -sSL https://install.python-poetry.org \
     | sed -e 's/allowed_executables = \["python", "python3"\]/allowed_executables = ["python3"]/' \
```
Comment (zyv) [inline bin/compile:40]: I have also removed this in #48, because this seems to not be necessary for the new installer anymore.

**Comment by felix11h** [comment]
Changes here have been merged in #48. Thanks again to everyone who contributed!
